### PR TITLE
Provide Pure metadata url override in Pure SDLC

### PR DESCRIPTION
**Pull request by gs-kotang**
#### What type of PR is this?
Improvement

#### What does this PR do / why is it needed ?

This change gives flexibility to choose a metadata url in Pure Sdlc. We use it to build Pure Metadata URL. We allow a list of allowed hosts to be specified in the server configuration, and we check that the specified override url is one of them.

#### Does this PR introduce a user-facing change?
No

### CODE REVIEW

File: legend-engine-language-pure-modelManager-sdlc/src/main/java/org/finos/legend/engine/language/pure/modelManager/sdlc/api/MetadataServerInfo.java [hunk 1]
```diff
@@ -35,12 +36,21 @@
 
     public MetadataServerInfo(MetaDataServerConfiguration metaDataServerConfiguration)
     {
+        String typeOfServerConnectionConfiguration = "";
+        String allowedOverrideUrls = "";
+        if (metaDataServerConfiguration.pure instanceof PureServerConnectionConfiguration)
+        {
+            typeOfServerConnectionConfiguration = "      \"_type\":\"pureServerConnectionConfiguration\"";
```
Comment (gs-ssh16) [inline legend-engine-language-pure-modelManager-sdlc/src/main/java/org/finos/legend/engine/language/pure/modelManager/sdlc/api/MetadataServerInfo.java:43]: Missing comma at the end?

File: legend-engine-pure-code-compiled-core/src/main/resources/core/pure/protocol/vX_X_X/invocations/execution.pure [hunk 2]
```diff
@@ -115,6 +115,11 @@ function meta::protocols::pure::vX_X_X::invocation::execution::execute::executeP
 }
 
 function meta::protocols::pure::vX_X_X::invocation::execution::execute::buildExecutionInput<T|m>(f:FunctionDefinition<{->T[m]}>[1], m:Mapping[1], runtime:Runtime[1],  exeCtx:ExecutionContext[0..1], version:String[1], serializationKind:String[1], executionMode:String[1], extensions:meta::pure::extension::Extension[*]):String[1]
+{
+  meta::protocols::pure::vX_X_X::invocation::execution::execute::buildExecutionInput($f, $m, $runtime, $exeCtx, $version, $serializationKind, $executionMode, $extensions, [])
+}
+
+function meta::protocols::pure::vX_X_X::invocation::execution::execute::buildExecutionInput<T|m>(f:FunctionDefinition<{->T[m]}>[1], m:Mapping[1], runtime:Runtime[1],  exeCtx:ExecutionContext[0..1], version:String[1], serializationKind:String[1], executionMode:String[1], extensions:meta::pure::extension::Extension[*], pureMetadataServer:String[0..1]):String[1]
```
Comment (gs-ssh16) [inline legend-engine-pure-code-compiled-core/src/main/resources/core/pure/protocol/vX_X_X/invocations/execution.pure:122]: Place metadata server param before extensions param

File: legend-engine-pure-code-compiled-core/src/main/resources/core/pure/protocol/vX_X_X/invocations/execution.pure [hunk 3]
```diff
@@ -152,6 +157,7 @@ function meta::protocols::pure::vX_X_X::invocation::execution::execute::buildExe
                                           _type = 'pure',
                                           baseVersion = '-1', //meta::vcs::svn::metamodel::getCurrentSvnRevisionNumber('/system')->toString(),
                                           version = 'none',
+                                          overrideUrl = $pureMetadataServer,
```
Comment (gs-ssh16) [inline legend-engine-pure-code-compiled-core/src/main/resources/core/pure/protocol/vX_X_X/invocations/execution.pure:160]: Also, can we have a simple unit test which asserts that override url is set

File: legend-engine-language-pure-modelManager-sdlc/src/main/java/org/finos/legend/engine/language/pure/modelManager/sdlc/pure/PureServerLoader.java [hunk 4]
```diff
@@ -44,14 +46,33 @@ public PureServerLoader(MetaDataServerConfiguration metaDataServerConfiguration)
...
 
-    public String buildPureMetadataVersionURL(String urlSuffix)
+    public String getBaseUrl(String overrideUrl)
     {
-        return metaDataServerConfiguration.getPure().getBaseUrl() + "/alloy/pureServerBaseVersion" + urlSuffix;
+        if (overrideUrl == null || overrideUrl.equals(metaDataServerConfiguration.getPure().getBaseUrl()))
+        {
+            return metaDataServerConfiguration.getPure().getBaseUrl();
+        }
+        List<String> allowedOverrideUrls = ((PureServerConnectionConfiguration)metaDataServerConfiguration.getPure()).allowedOverrideUrls;
```
Comment (gs-ssh16) [inline legend-engine-language-pure-modelManager-sdlc/src/main/java/org/finos/legend/engine/language/pure/modelManager/sdlc/pure/PureServerLoader.java:55]: Is this cast (PureServerConnectionConfiguration) backward compatible? I think we should check in above 'if' clause.

if (overrideUrl == null || overrideUrl.equals(metaDataServerConfiguration.getPure().getBaseUrl()) || !(metaDataServerConfiguration instanceof PureServerConnectionConfiguration))

Comment (gs-kotang) [reply legend-engine-language-pure-modelManager-sdlc/src/main/java/org/finos/legend/engine/language/pure/modelManager/sdlc/pure/PureServerLoader.java:55]: Added it. Providing a override url to the current version of metadata server configuration will now be a no-op.

File: legend-engine-pure-code-compiled-core/src/main/resources/core/pure/protocol/vX_X_X/invocations/execution.pure (hunk 3 above)
Comment (gs-kotang) [reply legend-engine-pure-code-compiled-core/src/main/resources/core/pure/protocol/vX_X_X/invocations/execution.pure:160]: Added

**Review by gs-ssh16 — APPROVED** [review]

**Pull request merged**